Assets/Scripts/Games/NipInTheBud/NipInTheBud.cs

```diff
@@ -168,26 +230,58 @@ public void BopToggle(double beat, float length, bool toggle, bool autoBop)
                 List<BeatAction.Action> bops = new List<BeatAction.Action>();
                 for (int i = 0; i < length; i++)
                 {
-                    bops.Add(new BeatAction.Action(beat + i, delegate { Bop(); }));
+                    bops.Add(new BeatAction.Action(beat + i, delegate { Bop(beat); }));
                 }
                 BeatAction.New(instance, bops);
             }
         }
 
-        public void Bop()
+        public void Bop(double beat)
         {
-            if (!noBop && !preparing && !queuePrepare && (Leilani.IsAnimationNotPlaying() || Leilani.IsPlayingAnimationNames("Idle"))) Leilani.DoScaledAnimationAsync("Bop", 0.5f, 0);
+            if (!noBop && !preparing && !queuePrepare && (Leilani.IsAnimationNotPlaying() || Leilani.IsPlayingAnimationNames("Idle")))
+            { Leilani.DoScaledAnimationAsync("Bop", 0.5f, 0);
+            Leilani.DoScaledAnimationAsync(bopExpression, 0.5f, 1);
+            
+            if ((bopExpression == "Happy" || bopExpression == "Sad") && !queueBopReset)
+                {
+                    ResetBopExpression(beat); //change beat to beat+1, remove beat+1 from ResetBopExpression
+                }
+            }
         }
 
-        public void QueueMosquito (double beat)
+        private ColorEase bgColorEase = new(defaultBGColor);
+
+        //call this in update
+        private void BackgroundColorUpdate()
+        {
+            bg.color = bgColorEase.GetColor();
+        }
+
+        public void BackgroundColor(double beat, float length, Color startColor, Color endColor, int ease)
+        {
+            bgColorEase = new(beat, length, startColor, endColor, ease);
+        }
+
+        private void PersistColor(double beat)
+        {
+            var allEventsBeforeBeat = EventCaller.GetAllInGameManagerList("nipInTheBud", new string[] { "fade background" }).FindAll(x => x.beat < beat);
+            if (allEventsBeforeBeat.Count > 0)
+            {
+                allEventsBeforeBeat.Sort((x, y) => x.beat.CompareTo(y.beat)); //just in case
+                var lastEvent = allEventsBeforeBeat[^1];
```

Comment:
instead of FindAll, use FindLast. it spends less time finding objects and doesn't create a new list. the sorting is also redundant, as whenever events can be used, they're sorted. if they aren't, we have an even bigger issue.